### PR TITLE
netmaster listen and control port fix

### PR DESCRIPTION
**Pull request by DivyaVavili**
Changes included:
- introduce a control_url cmd line arg for netmaster for control pkts
- listen_url is used for any external communication

If no listen_url/control_url is specified, netmaster listens on :9999 by default
If no IP is mentioned in control_url, local_IP:control_url is used for netplugin->netmaster control communication

### CODE REVIEW

File: netmaster/daemon/daemon.go [hunk 1]
```diff
@@ -88,18 +91,26 @@ func (d *MasterDaemon) Init() {
...
-	localIP, err := getLocalAddr()
-	if err != nil {
-		log.Fatalf("Error getting local IP address. Err: %v", err)
+	var err error
+
+	ctrlURL := strings.Split(d.ControlURL, ":")
+	masterIP := ctrlURL[0]
+	masterPort, _ := strconv.Atoi(ctrlURL[1])
+
+	if len(masterIP) == 0 {
```
Comment (jojimt) [inline netmaster/daemon/daemon.go]: Would be nice to perform this check when we parse this command line and init d.ControlURL there.

File: netmaster/daemon/daemon.go [hunk 2]
```diff
@@ -246,28 +260,44 @@ func (d *MasterDaemon) runLeader() {
...
 		log.Fatalln(err)
 	}
+	log.Infof("Netmaster listening on %s for control packets", d.ControlURL)
+	ctrlListener = utils.ListenWrapper(ctrlListener)
 
-	log.Infof("Netmaster listening on %s", d.ListenURL)
+	go server.Serve(ctrlListener)
 
-	listener = utils.ListenWrapper(listener)
+	if !((len(d.ControlURL) == len(d.ListenURL)) && strings.Contains(d.ControlURL, d.ListenURL)) {
```
Comment (jojimt) [inline netmaster/daemon/daemon.go]: Better to use strings.Compare

File: netmaster/daemon/daemon.go [hunk 3]
```diff
@@ -278,15 +308,25 @@ func (d *MasterDaemon) runFollower() {
...
 		log.Fatalln(err)
 	}
+	ctrlListener = utils.ListenWrapper(ctrlListener)
 
-	listener = utils.ListenWrapper(listener)
+	go server.Serve(ctrlListener)
 
-	// start server
-	go server.Serve(listener)
+	if !((len(d.ControlURL) == len(d.ListenURL)) && strings.Contains(d.ControlURL, d.ListenURL)) {
```
Comment (jojimt) [inline netmaster/daemon/daemon.go]: strings.Compare

File: netmaster/daemon/daemon.go [hunk 4]
```diff
@@ -329,14 +372,22 @@ func (d *MasterDaemon) becomeFollower() {
 func (d *MasterDaemon) RunMasterFsm() {
 	var err error
 
-	// Get the address to be used for local communication
-	localIP, err := getLocalAddr()
-	if err != nil {
-		log.Fatalf("Error getting local IP address. Err: %v", err)
+	masterURL := strings.Split(d.ControlURL, ":")
+	masterIP, masterPort := masterURL[0], masterURL[1]
+	if len(masterIP) == 0 {
```
Comment (jojimt) [inline netmaster/daemon/daemon.go]: Move to cmdline parsing.

File: netmaster/daemon/daemon.go [hunk 5]
```diff
@@ -246,28 +260,44 @@ func (d *MasterDaemon) runLeader() {
 	// Create HTTP server and listener
 	server := &http.Server{Handler: router}
 	server.SetKeepAlivesEnabled(false)
-	listener, err := net.Listen("tcp", d.ListenURL)
+
+	ctrlListener, err = net.Listen("tcp", d.ControlURL)
```
Comment (jojimt) [inline netmaster/daemon/daemon.go]: We need to handle the scenario where controlURL = "192.1.1.1:9999" and listernURL = ":9999".

File: netmaster/main.go [hunk 6]
```diff
@@ -62,10 +63,14 @@ func parseOpts(opts *cliOpts) error {
 		"cluster-store",
 		"etcd://127.0.0.1:2379",
 		"Etcd or Consul cluster store url.")
+	flagSet.StringVar(&opts.controlURL,
+		"control-url",
```
Comment (jojimt) [inline netmaster/main.go]: Shall we call it internal-url instead of control-url?

File: test/systemtests/basic_test.go [hunk 7]
```diff
@@ -223,3 +224,62 @@ func (s *systemtestSuite) testBasicSvcDiscovery(c *C, encap string) {
...
+		tenant, err := cliClient.TenantGet("default")
+		if err != nil || strings.Contains(tenant.TenantName, "default") {
+			err = fmt.Errorf("Client request to %s failed. Err: %+v", clientURL, err)
+		}
+		c.Assert(err, IsNil)
+
+		c.Assert(masterNode.stopNetmaster(), IsNil)
+		time.Sleep(5 * time.Second)
+		masterPort = "888" + masterIP[:len(masterIP)-1]
+		c.Assert(masterNode.startNetmaster(fmt.Sprintf("--control-url=:%s --listen-url=:%s", masterPort, masterPort)), IsNil)
```
Comment (jojimt) [inline test/systemtests/basic_test.go]: Pls cover the case of control-url=a.b.c.d:9999 and listen-url=:9999.

File: netmaster/main.go (hunk 6 above)
Comment (DivyaVavili) [inline netmaster/main.go]: Specific reason for the suggestion? I think control-url suggests using this for exchanging control protocol information. If you have a strong UX sense for naming it as internal-url, we can discuss and make the change.

File: test/systemtests/basic_test.go (hunk 7 above)
Comment (DivyaVavili) [inline test/systemtests/basic_test.go]: Will add the case

File: netmaster/daemon/daemon.go (hunk 5 above)
Comment (DivyaVavili) [inline netmaster/daemon/daemon.go]: Will make the necessary change

**Comment by DivyaVavili** [comment]
@srampal: This PR is a fix for #705. If by debug port, you mean the port mentioned in --listen-url argument; yes, this PR will address it.

File: scripts/python/api/tnode.py [hunk 8]
```diff
@@ -102,7 +102,7 @@ def startNetplugin(self, args=""):
     # Start netmaster process
     def startNetmaster(self):
         ssh_object = self.sshConnect(self.username, self.password)
-        command = "GOPATH=/opt/gopath " + self.binpath + "/netmaster -cluster-store " + os.environ["CONTIV_CLUSTER_STORE"] + " > /tmp/netmaster.log 2>&1"
+        command = "GOPATH=/opt/gopath " + self.binpath + "/netmaster --listen-url=netmaster:" + os.environ["CONTIV_NETMASTER_PORT"] + " -cluster-store " + os.environ["CONTIV_CLUSTER_STORE"] + " > /tmp/netmaster.log 2>&1"
```
Comment (gkvijay) [inline scripts/python/api/tnode.py]: Are we making listen-url mandatory. Can't we use the default option

File: netmaster/main.go [hunk 9]
```diff
@@ -113,6 +118,7 @@ func main() {
 	// create master daemon
 	d := &daemon.MasterDaemon{
 		ListenURL:    opts.listenURL,
+		ControlURL:   opts.controlURL,
```
Comment (gkvijay) [inline netmaster/main.go]: Can you parse the listenURL and controlURL and check if its valid format

**Comment by jojimt** [comment]
No the issue @srampal is referring to is separate.

File: netmaster/daemon/daemon.go (hunk 1 above)
Comment (DivyaVavili) [inline netmaster/daemon/daemon.go]: Moved the sanity checks and defaults allocation to command parsing phase

File: netmaster/daemon/daemon.go [hunk 10]
```diff
@@ -88,18 +91,26 @@ func (d *MasterDaemon) Init() {
 }
 
 func (d *MasterDaemon) registerService() {
-	// Get the address to be used for local communication
-	localIP, err := getLocalAddr()
-	if err != nil {
-		log.Fatalf("Error getting local IP address. Err: %v", err)
+	var err error
+
+	ctrlURL := strings.Split(d.ControlURL, ":")
```
Comment (DivyaVavili) [inline netmaster/daemon/daemon.go]: I've made changes to do sanity checking during cmdline parsing

File: scripts/python/api/tnode.py (hunk 8 above)
Comment (DivyaVavili) [inline scripts/python/api/tnode.py]: Made the necessary changes

File: netmaster/main.go (hunk 9 above)
Comment (DivyaVavili) [inline netmaster/main.go]: handled during cmdline parsing

**Comment by DivyaVavili** [comment]
@jojimt @gkvijay: PTAL. I've made some changes to how the listener ports are sanitized and listeners are initialized. Please review them. I've separated the changes to logical commits for easy review.

File: netmaster/daemon/daemon.go [hunk 11]
```diff
@@ -271,33 +252,53 @@ func (d *MasterDaemon) runFollower() {
...
-	// Register netmaster service
-	d.registerService()
+	listenURL := strings.Split(d.ListenURL, ":")
+	controlURL := strings.Split(d.ControlURL, ":")
 
-	// just wait on stop channel
-	log.Infof("Listening in follower mode")
-	<-d.stopFollowerChan
+	if strings.Compare(listenURL[1], controlURL[1]) != 0 {
+		ctrlListener, err := net.Listen("tcp", d.ControlURL)
```
Comment (jojimt) [inline netmaster/daemon/daemon.go]: I think we have to do this for listernURL[0] != 0.0.0.0 && listenURL[0] != controlURL[0] as well.

File: netmaster/main.go (hunk 6 above)
Comment (jojimt) [inline netmaster/main.go]: I think it's fine. We can leave it as control-url

File: netmaster/main.go [hunk 12]
```diff
@@ -97,6 +110,51 @@ func execOpts(opts *cliOpts) {
...
 	}
 
+	// Validate listen and control URL options
+	listenURL := strings.Split(opts.listenURL, ":")
+	controlURL := strings.Split(opts.controlURL, ":")
+	if len(listenURL) != 2 {
+		log.Fatalf("Listen URL not in proper format. Valid format: [IP]:Port")
+	}
+	if len(controlURL) != 2 {
+		log.Fatalf("Control URL not in proper format. Valid format: IP:Port")
```
Comment (jojimt) [inline netmaster/main.go]: Can we also make sure we don't accept 0.0.0.0 for controlURL[0] because without an IP, controlURL does not server its purpose.

**Comment by DivyaVavili** [comment]
@jojimt: Handled the cases mentioned. PTAL

**Pull request merged**